lib/rgb_matrix/engine.ex

```diff
@@ -96,64 +100,76 @@ defmodule RGBMatrix.Engine do
     %State{state | paintables: paintables}
   end
 
-  defp set_animation(state, animation) do
+  defp set_animation(state, animation_type) do
+    {render_in, animation} = Animation.new(animation_type, state.leds)
+
+    state = schedule_next_render(state, render_in)
+
     %State{state | animation: animation}
   end
 
-  @impl GenServer
-  def handle_info(:get_next_frame, state) do
-    animation = Animation.next_frame(state.animation)
+  defp schedule_next_render(state, :ignore) do
+    state
+  end
 
-    state.paintables
-    |> Map.values()
-    |> Enum.each(fn paint_fn ->
-      paint_fn.(animation.next_frame)
-    end)
+  defp schedule_next_render(state, :never) do
+    cancel_timer(state)
+  end
 
-    Process.send_after(self(), :get_next_frame, animation.delay_ms)
-    {:noreply, set_animation(state, animation)}
+  defp schedule_next_render(state, 0) do
+    send(self(), :render)
+    cancel_timer(state)
   end
 
-  @impl GenServer
-  def handle_info({:reset_animation, reset_animation}, state) do
-    {:noreply, set_animation(state, reset_animation)}
+  defp schedule_next_render(state, ms) when is_integer(ms) and ms > 0 do
+    state = cancel_timer(state)
+    %{state | timer: Process.send_after(self(), :render, ms)}
   end
 
-  @impl GenServer
-  def handle_info({:reply, from, reset_animation}, state) do
-    GenServer.reply(from, :ok)
+  defp cancel_timer(%{timer: nil} = state), do: state
 
-    {:noreply, set_animation(state, reset_animation)}
+  defp cancel_timer(state) do
+    Process.cancel_timer(state.timer)
+    %{state | timer: nil}
   end
 
-  @impl GenServer
-  def handle_cast({:play_animation, %{loop: loop} = animation}, state)
-      when is_integer(loop) and loop >= 1 do
-    current_animation = state.animation
-    expected_duration = Animation.duration(animation)
-    Process.send_after(self(), {:reset_animation, current_animation}, expected_duration)
+  @impl true
+  def handle_info(:render, state) do
+    {new_colors, render_in, animation} = Animation.render(state.animation)
```

Comment:
```suggestion
    {render_in, new_colors, animation} = Animation.render(state.animation)
```
To match my suggestion in `animation.ex`.